ecl/hql/hqlgram2.cpp

```diff
@@ -11022,107 +11146,111 @@ void HqlGram::syntaxError(const char *s, int token, int *expected)
     if (yytext)
         column -= strlen(yytext);
 
-    if (expected)
-        simplifyExpected(expected);
-
     StringBuffer msg;
     if (expected == NULL) // expected is NULL when fatal internal error occurs.
     {
         msg.append(s);
-    } 
+        reportError(ERR_EXPECTED, msg.str(), lineno, column, pos);
+        return;
+    }
     else if (token==UNKNOWN_ID)
     {
         msg.append("Unknown identifier");
         if (yytext && *yytext)
+        {
             msg.append(" \"").append(yytext).append('\"');
-
+            syntaxErrorCompareExpected(msg, yytext, expected, expectedLength);
+        }
         reportError(ERR_UNKNOWN_IDENTIFIER,msg.str(), lineno, column, pos);
         return;
     }
     else if ((token == '.') && (expected[0] == ASSIGN) && !expected[1])
     {
-        reportError(ERR_UNKNOWN_IDENTIFIER,"Unknown identifier before \".\" (expected :=)", lineno, column, pos);
+        /*Ambiguous situation - := as the only expected token only occurs after an unknown_id (need to check this still)
+         *and thus assuming that the unknown_id was itentional it can be inferred that the '.' can't be as this needs a
+         *known_id. Alternatively, it can be assumed that the '.' was intentional and thus
+         *that the unknown_id is the error and should actually have been a known_id.
+         **/
+        reportError(ERR_UNKNOWN_IDENTIFIER,"Either unknown identifier before \".\" or \".\" not intended : expected(only) :=", lineno, column, pos);
```

Comment:
Personally, I have always found the existing form confusing as it is ambiguous and feel that others could also do so. Perhaps I have misunderstood something. I'll try to explain my thinking.
Looking at the condition that triggers this error all we know (without looking further) is that a '.' was given as an invalid next token thus generating the syntax error and that also the only allowed token for this state is ':=', the assign op. Lets reduce the context further and assume that the inclusion of 'unknown id' in the message is due to the fact that the only place in the state table that has a single assign op as the expected token is after an unknown id which makes sense if you know the language a little. Ok, so we now have an unknown id followed by a '.' where ':=' was expected instead. The issue is how do we construct an error message that covers as many of the users intended scenarios. This is the point, right, they have made an error but what was their intention. In this example we can have two scenarios:
1) (intention) {code}unknown_id := {code} (actual) {code}unknown_id. :={code}
2) (intention) {code}known_id '.'{code} (actual) {code} unknown_id .{code}
and the question is which is the error the identifier being either a known or unknown id when it should be the other or that the '.' is present when it shouldn't be? The user may have intended for a known id but made a typo resulting in this being parsed as an unknown id in which a '.' cannot follow and thus it is the unknown id that is the source of the error (which is what the current form suggests). Alternatively, the user could have intended for an unknown id and the '.' is the typo/error. The user may not even be aware of these two scenarios. Furthermore, it is confusing to which token the expected token referes to, the id or the '.'. 
I don't believe this messaeg is less clear than it was but I would still say that it is still not as clear as I would like it. Maybe this:
Either unintended unknown identifier before '.' or '.' not intended : expected ':=' instead of '.'